### PR TITLE
In app message template

### PR DESCRIPTION
**Pull request by MattLichtenstein**
## Description

Implements the messages template view for addon in-app messages

## Reference

[VPN-2489: Implement messages template view](https://mozilla-hub.atlassian.net/browse/VPN-2489)

## Checklist
    
- [ ] My code follows the style guidelines for this project
- [ ] I have not added any packages that contain high risk or unknown licenses (GPL,  LGPL, MPL, etc. consult with DevOps if in question)
- [ ] I have performed a self review of my own code
- [ ] I have commented my code PARTICULARLY in hard to understand areas
- [ ] I have added thorough tests where needed

### CODE REVIEW

File: addons/message_survey_staging_4s/manifest.json [hunk 1]
```diff
@@ -2,7 +2,6 @@
   "api_version": "0.1",
   "id": "message_survey_staging_4s",
   "name": "Staging Survey 4 secs",
-  "translatable": false,
```
Comment (flodolo) [inline addons/message_survey_staging_4s/manifest.json:5]: This too, there's placeholder content (line 20-25)

**Review by flodolo — CHANGES_REQUESTED** [review]

File: addons/message_survey_staging_locale/manifest.json [hunk 2]
```diff
@@ -2,7 +2,6 @@
   "api_version": "0.1",
   "id": "message_survey_staging_locale",
   "name": "Staging Survey 1 secs",
-  "translatable": false,
```
[the remark this answers is not in the document]
Comment (MattLichtenstein) [reply addons/message_survey_staging_locale/manifest.json:5]: Ahh, this was just a bad rebase 😅 thanks for pointing it out

File: addons/message_survey_staging_4s/manifest.json [hunk 3]
```diff
@@ -12,7 +12,8 @@
     "id": "message_survey_staging_4s",
```
Comment (lesleyjanenorton) [inline addons/message_survey_staging_4s/manifest.json:12]: same content/nomenclature nit here from #4309. Could be a follow up.

**Review by lesleyjanenorton — CHANGES_REQUESTED** [review]
Looking good!

File: translations/strings.yaml [hunk 4]
```diff
@@ -697,6 +697,21 @@ inAppMessaging:
   deleteMessage:
     value: Delete message
     comment: An action to delete a message from the users inbox
+  warningBadge:
+    value: Warning
+    comment: A badge shown in a message signifying that this is a warning message
```
Comment (flodolo) [inline translations/strings.yaml:702]: Are these actual text badges or alt text for images/icons?

Comment (MattLichtenstein) [reply translations/strings.yaml:702]: They are badges with text:

Screen Shot 2022-09-04 at 1 35 57 PM
Screen Shot 2022-09-04 at 1 36 35 PM

Comment (flodolo) [reply translations/strings.yaml:702]: How much space is there? For example, "New update" would be "Nuovo aggiornamento" in Italian (in French update is `Mise à jour`, in Spanish `Actualización`).

File: addons/message_survey_staging_10s/manifest.json [hunk 5]
```diff
@@ -12,8 +12,21 @@
...
+      "type": "button",
+      "style": "primary",
+      "content": "Renew subscription",
+      "javascript": "PresentSubscriptionManagementScreen.js"
+      },
+      { "id": "c_2",
+        "type": "button",
+        "style": "link",
+        "content": "Get help",
+        "javascript": "PresentGetHelpScreen.js"
```
[the remark this answers is not in the document]
Comment (bakulf) [reply addons/message_survey_staging_10s/manifest.json:28]: can you revert all of these changes? And file new messages for staging if you want to test things.
Asking this because this survey has been renamed and I guess you will have a few conflict to fix.
I also would like to make this add-on a survey one: label, text, etc.

File: addons/message_survey_staging_4s/manifest.json [hunk 6]
```diff
@@ -28,6 +29,18 @@
       { "id": "c_2",
         "type": "text",
         "content": "You can update instantly by hitting the update button below!" 
+      },
+      { "id": "c_3",
+      "type": "button",
+      "style": "primary",
+      "content": "Update now",
+      "javascript": "PresentUpdateScreen.js"
```
Comment (bakulf) [inline addons/message_survey_staging_4s/manifest.json:37]: where are these files?

File: nebula/ui/components/VPNComposerView.qml [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+            }
+
+
+            function getSourceComponent() {
+                if (modelData instanceof VPNComposerBlockTitle) return titleBlock
+                if (modelData instanceof VPNComposerBlockText) return textBlock
+                if (modelData instanceof VPNComposerBlockOrderedList ||
+                    modelData instanceof VPNComposerBlockUnorderedList) return listBlock
+                if (modelData instanceof VPNComposerBlockButton)
+                    return composerBlock.style === VPNComposerBlockButton.Link ? linkButtonBlock : buttonBlock
```
Comment (bakulf) [inline nebula/ui/components/VPNComposerView.qml:55]: what about if here we return an error somehow? At least a `console.error()`

File: src/addons/addonmessage.cpp [hunk 8]
```diff
@@ -224,3 +226,21 @@ qint64 AddonMessage::planDateRetranslationInternal(
...
+    } else if (badge == "critical") {
+      m_badge = Critical;
+    } else if (badge == "new_update") {
+        m_badge = NewUpdate;
+    } else if (badge == "whats_new") {
+        m_badge = WhatsNew;
+    } else if (badge == "survey") {
+        m_badge = Survey;
+    } else {
+      logger.error() << "Unsupported badge type" << badge;
```
Comment (bakulf) [inline src/addons/addonmessage.cpp:244]: mmm do we want to return an error here and reject the addon?

File: src/commands/commandui.cpp [hunk 9]
```diff
@@ -523,6 +523,7 @@ int CommandUI::run(QStringList& tokens) {
           return obj;
         });
 
+
```
Comment (bakulf) [inline src/commands/commandui.cpp:526]: revert this change :)

File: src/composer/composerblock.cpp [hunk 10]
```diff
@@ -9,6 +9,7 @@
 #include "composerblocktitle.h"
 #include "composerblockorderedlist.h"
 #include "composerblockunorderedlist.h"
+#include "composerblockbutton.h"
```
Comment (bakulf) [inline src/composer/composerblock.cpp:12]: can you sort all of these headers? (with composerblock.h as the first one)

**Review by bakulf — CHANGES_REQUESTED** [review]

File: nebula/ui/components/VPNComposerView.qml [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+            }
+
+            function getTopMargin() {
+                if (modelData instanceof VPNComposerBlockTitle) {
+                    switch(composer.view) {
+                    case VPNComposerView.View.Guide:
+                    case VPNComposerView.View.Message:
```
[the remark this answers is not in the document]
Comment (MattLichtenstein) [reply nebula/ui/components/VPNComposerView.qml:62]: Nope, this was just to set a precedent in case we use the composer somewhere else in the future that does have a different value - I'll fix this

File: nebula/ui/components/VPNComposerView.qml [hunk 12]
```diff
@@ -0,0 +1,178 @@
...
+import QtQuick 2.15
+import QtQuick.Layouts 1.15
+import Qt.labs.qmlmodels
+
+import Mozilla.VPN 1.0
+import Mozilla.VPN.qmlcomponents 1.0
+import components 0.1
+
+ColumnLayout {
+    id: composer
```
[the remark this answers is not in the document]
Comment (MattLichtenstein) [reply nebula/ui/components/VPNComposerView.qml:14]: Which platform? It's working for me on mac

File: nebula/ui/components/VPNComposerView.qml (hunk 7 above)
Comment (MattLichtenstein) [reply nebula/ui/components/VPNComposerView.qml:55]: I'd lean towards not showing the message if one of the blocks doesn't load... We'd have to come up with some UI/UX error state for that case since them message would still be visible in the inbox because we wouldn't perform this check until the message has been clicked. Putting a console log for now

File: src/addons/addonmessage.cpp (hunk 8 above)
Comment (MattLichtenstein) [reply src/addons/addonmessage.cpp:244]: I feel like the badge isn't mission critical to the addon and the message would still be worth showing, but I could go either way

File: translations/strings.yaml [hunk 13]
```diff
@@ -688,7 +688,7 @@ tutorialPopup:
   tutorialLeavePrimaryButtonLabel: Resume tutorial
 
 inAppMessaging:
-  menuTitle: Inbox
+  menuTitle: Messages
```
Comment (flodolo) [inline translations/strings.yaml:691]: note: this is OK, since this hasn't been translated yet
https://github.com/mozilla-l10n/mozilla-vpn-client-l10n/pull/289/files#diff-23c0d0ff5a7678a6d691121f6e2a4298bd8bbabb5e79f9b5049c14a880cd38e7

File: translations/strings.yaml (hunk 4 above)
Comment (MattLichtenstein) [reply translations/strings.yaml:702]: There should be plenty of space for those examples

 Screen Shot 2022-09-06 at 2 24 03 PM

But if the string gets really long (which is technically possible), this would be the current UI - but I am looking for ways to improve that

Screen Shot 2022-09-06 at 2 25 29 PM

Comment (flodolo) [reply translations/strings.yaml:702]: Thanks, wasn't sure if they would show up near the title, not on a separate line.

**Review by flodolo — APPROVED** [review]

File: nebula/ui/components/VPNComposerView.qml (hunk 12 above)
[the remark this answers is not in the document]
Comment (lesleyjanenorton) [reply nebula/ui/components/VPNComposerView.qml:14]: interesting! I'm also on a mac. Qt version is 6.2.4, debug build built with XCode.

File: scripts/ci/jsonSchemas/composer_block.json [hunk 14]
```diff
@@ -47,6 +47,14 @@
...
       "type": "string",
       "description": "A type for this block: text, title, ulist, olist, etc"
+    },
+    "style": {
+      "type": "string",
+      "description": "A style required for button composer blocks"
+    },
+    "javascript": {
+      "type": "string",
+      "description": "A javascript file that performs an action for button compsoer blocks"
```
Comment (MattLichtenstein) [inline scripts/ci/jsonSchemas/composer_block.json:57]: @bakulf is there anything I should do here to specify that this is specifically for button composer blocks? Should I create new schema for buttons?

File: translations/strings.yaml (hunk 4 above)
Comment (MattLichtenstein) [reply translations/strings.yaml:702]: Badge text is now elided 

Screen Shot 2022-09-06 at 5 55 10 PM

File: addons/message_expiring_subscription/manifest.json [hunk 15]
```diff
@@ -0,0 +1,36 @@
+{
+  "api_version": "0.1",
+  "id": "message_expiring_subscription",
+  "name": "Staging expiring subscription message",
+  "type": "message",
+  "translatable": false,
+  "conditions": {
+    "env": "staging",
+    "trigger_time": 1
```
Comment (bakulf) [inline addons/message_expiring_subscription/manifest.json:9]: We can remove this line

File: addons/message_update/manifest.json [hunk 16]
```diff
@@ -0,0 +1,51 @@
+{
+  "api_version": "0.1",
+  "id": "message_update",
+  "name": "Staging update message",
+  "translatable": false,
+  "type": "message",
+  "conditions": {
+    "env": "staging",
+    "trigger_time": 1
```
Comment (bakulf) [inline addons/message_update/manifest.json:9]: Remove this

File: nebula/ui/components/VPNComposerView.qml [hunk 17]
```diff
@@ -0,0 +1,169 @@
...
+                    horizontalAlignment: Text.AlignLeft
+                }
+            }
+
+            Component {
+                id: listBlock
+
+                VPNInterLabel {
+                    property string listType: loader.composerBlock && composerBlock.type === "olist" ? "ol" : "ul"
+                    property var tagsList: loader.composerBlock.subBlocks.map(subBlock => `<li>${subBlock}</li>`)
```
Comment (bakulf) [inline nebula/ui/components/VPNComposerView.qml:125]: For now it's not an issue but eventually we need to be sure that `subBlock` does not contain < or >. Or we need to escape a few chars.

File: nebula/ui/components/VPNComposerView.qml [hunk 18]
```diff
@@ -0,0 +1,169 @@
...
+            Component {
+                id: buttonBlock
+
+                VPNButton {
+                    anchors.left: parent.left
+                    anchors.right: parent.right
+
+                    implicitHeight: VPNTheme.theme.rowHeight
+                    text: loader.composerBlock.text
+                    colorScheme: loader.composerBlock.style === VPNComposerBlockButton.Primary ? VPNTheme.theme.blueButton : VPNTheme.theme.redButton
```
Comment (bakulf) [inline nebula/ui/components/VPNComposerView.qml:149]: What about the other button types? Should we remove the unsupported types?

**Review by bakulf — APPROVED** [review]

File: nebula/ui/components/VPNComposerView.qml (hunk 18 above)
Comment (MattLichtenstein) [reply nebula/ui/components/VPNComposerView.qml:149]: What other button types? We only have primary, destructive, and link. Primary and destructive are handled here and link is a different component.

File: addons/message_update/PresentUpdateScreen.js [hunk 19]
```diff
@@ -0,0 +1,3 @@
+(function(api) {
+  api.navigator.requestScreen(api.navigator.ScreenUpdateRequired);
```
Comment (bakulf) [inline addons/message_update/PresentUpdateScreen.js:2]: `ScreenUpdateRecommended`

**Review by lesleyjanenorton — APPROVED** [review]
